### ISSUE TITLE
Reading `character(0)` objects does not work

### ISSUE THREAD

**Issue by rcannood** [hhoeflin/hdf5r#118]
Hello there! Thank you for this super useful package!

I'm having problems when I save a `character(0)` object, and then try to read it from h5.
```R
library(hdf5r)

file.remove("~/test.h5")
file_h5 <- H5File$new("~/test.h5", mode = "w")
file_h5[["numeric"]] <- numeric(0)
file_h5[["character"]] <- character(0)
```
When reading it, I get the following error message:
```
> file_h5[["numeric"]]
Class: H5D
Dataset: /numeric
Filename: /home/rcannood/test.h5
Access type: H5F_ACC_RDWR
Datatype: H5T_IEEE_F64LE
Space: Type=Simple     Dims=0     Maxdims=Inf
Chunk: 1024
> file_h5[["numeric"]][]
numeric(0)
> file_h5[["character"]]
Class: H5D
Dataset: /character
Filename: /home/rcannood/test.h5
Access type: H5F_ACC_RDWR
Datatype: H5T_STRING {
      STRSIZE H5T_VARIABLE;
      STRPAD H5T_STR_NULLTERM;
      CSET H5T_CSET_ASCII;
      CTYPE H5T_C_S1;
   }
Space: Type=Simple     Dims=0     Maxdims=Inf
Chunk: 1024
> file_h5[["character"]][]
Error in self$read_low_level(file_space = self_space_id, mem_space = mem_space_id,  : 
  HDF5-API Errors:
    error #000: ../../src/H5D.c in H5Dvlen_reclaim(): line 830: invalid argument
        class: HDF5
        major: Invalid arguments to routine
        minor: Bad value
```

Is this a bug? Can I circumvent this problem?

Thanks,
Robrecht

**Comment by hhoeflin** [comment]
Hi,

thanks for reporting this bug. Yes, it is one. The reason is probably that 'character(0)' doesn't define the datatype well. For strings, the length of the string has to be estimated, and for 'character(0)' that is not possible.

Best way to work around this issue is to use the 'create_dataset' function and explicitly ask for the desired datatype and size for a set, rather than through an example R object.

**Comment by rcannood** [comment]
Hi @hhoeflin,

Thanks for your suggestion. Is this how you suggest I try to solve it?
```R
file_h5$create_dataset(
  "character2",
  dtype = h5types$char,
  space = H5S$new(
    "simple", 
    dims = 0, 
    maxdims = Inf),
  chunk_dims = 1024
)
```
Unfortunately, this results in the same error:
```
> file_h5[["character2"]]
Class: H5D
Dataset: /character2
Filename: /home/rcannood/test.h5
Access type: H5F_ACC_RDWR
Datatype: H5T_STRING {
      STRSIZE H5T_VARIABLE;
      STRPAD H5T_STR_NULLTERM;
      CSET H5T_CSET_ASCII;
      CTYPE H5T_C_S1;
   }
Space: Type=Simple     Dims=0     Maxdims=Inf
Chunk: 1024
> file_h5[["character2"]][]  
Error in self$read_low_level(file_space = self_space_id, mem_space = mem_space_id,  : 
  HDF5-API Errors:
    error #000: ../../src/H5D.c in H5Dvlen_reclaim(): line 830: invalid argument
        class: HDF5
        major: Invalid arguments to routine
        minor: Bad value
```

**Comment by rcannood** [comment]
This problem is solved in hdf5r 1.3.11! :partying_face: 

```r
> library(hdf5r)
> 
> file_h5 <- H5File$new("~/test.h5", mode = "w")
> file_h5[["numeric"]] <- numeric(0)
> file_h5[["character"]] <- character(0)
> 
> file_h5[["numeric"]][]
numeric(0)
> file_h5[["character"]][]
character(0)
```
